## Model Training and Comparison

In [1]:
data = pd.read_csv("feature_engineered_data.csv")
data.head()

<IPython.core.display.Javascript object>

Average_Salary  Years_Experience  Education_Level  AI_Exposure_Index  \
0       -1.259374          1.524586                2          -1.131450   
1        1.271092          0.609079                3           0.418080   
2        1.642772         -1.450811                1           1.263278   
3        1.362849         -0.191989                3          -0.391902   
4       -0.548382          0.837956                1           0.065914   

   Tech_Growth_Factor  Automation_Probability_2030  Risk_Category  Skill_1  \
0            0.989695                         0.85              2     0.45   
1            0.398639                         0.05              0     0.02   
2            0.642015                         0.81              2     0.01   
3           -1.096386                         0.60              1     0.43   
4            1.615519                         0.64              1     0.75   

   Skill_2  Skill_3  ...  Job_Title_Security Guard  \
0     0.10     0.46  ...                         1   
1     0.52     0.40  ...                         0   
2     0.94     0.56  ...                         0   
3     0.21     0.57  ...                         0   
4     0.54     0.59  ...                         0   

   Job_Title_Software Engineer  Job_Title_Teacher  Job_Title_Truck Driver  \
0                            0                  0                       0   
1                            0                  0                       0   
2                            0                  0                       0   
3                            1                  0                       0   
4                            0                  0                       0   

   Job_Title_UX Researcher  total_skill_score  Ai_risk_score  job_Stablity  \
0                        0               3.85      -0.338870      2.121145   
1                        0               4.26       0.270848     -0.019441   
2                        0               4.51       1.081967     -0.621263   
3                        0               4.92       0.004859     -0.704485   
4                        0               4.65       0.295548      1.549606   

   exp_salary  skill_eff  
0   -1.920024   1.525003  
1    0.774195   2.647477  
2   -2.383352 -10.004194  
3   -0.261652   6.089029  
4   -0.459519   2.529985  

[5 rows x 41 columns]

## 1. LINEAR REGRESSION

In [3]:
features_to_drop = [
    'Average_Salary'           # contains salary → leakage
]

X = data.drop(columns=features_to_drop)
y = data['Average_Salary']

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [5]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(X_train, y_train)
lr_pred = lr.predict(X_test)


In [6]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators=300, max_depth=10, random_state=42)
rf.fit(X_train,y_train)
rf_pred = rf.predict(X_test)


In [7]:
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score

xgb = XGBRegressor(n_estimators = 300, learning_rate = 0.05, max_depth = 6, subsample = 0.8, colsample_bytree = 0.8,
                  random_state=42)
xgb.fit(X_train, y_train)
xgb_pred = xgb.predict(X_test)

In [8]:
## creation of function to evaluate models

def evaluate_model(name, y_test, pred):
    mse = mean_squared_error(y_test, pred)
    rmse = mse**0.5
    r2 = r2_score(y_test, pred)
    return [name, mse, rmse, r2]

In [9]:
results = []

results.append(evaluate_model("Linear Regression", y_test, lr_pred))
results.append(evaluate_model("Random Forest", y_test, rf_pred))
results.append(evaluate_model("XGBoost", y_test, xgb_pred))

comparison_df = pd.DataFrame(results, columns=["Model", "MSE", "RMSE","R2 Score"])
comparison_df

<IPython.core.display.Javascript object>

Model       MSE      RMSE  R2 Score
0  Linear Regression  1.023032  1.011451 -0.005046
1      Random Forest  0.196506  0.443290  0.806949
2            XGBoost  0.130926  0.361838  0.871375

### CONCLUSION

### Which Model should you choose?

1. XGBoost: Best Overall
   - Highest R2 ~ 0.87
   - Best generalisation
   - Handle Non-linear values
2. Random Forest
   - Strong and Interpretable
3. Linear Regression
   - Good base but can't handle non-linear values